src/emmy/tape.cljc

```diff
@@ -246,24 +305,104 @@
     (recur (tape-primal v))
     v))
 
-;; ### Reverse-pass support
-
-(defn ^:no-doc topological-sort-by-id
+;; ### Comparison, Control Flow
+;;
+;; Functions like `=`, `<` and friends don't have derivatives; instead, they're
+;; used for control flow inside of Clojure functions. To play nicely with these
+;; functions, the [[TapeCell]] API exposes a number of methods for comparing
+;; numbers on ONLY their finite parts.
+;;
+;; Why? If `x` is a [[TapeCell]] instance, `(< x 10)` needs to return true
+;; whenever a non-[[TapeCell]] `x` would return true. To make this work, these
+;; operations look only at the [[tape-primal]].
+;;
+;; If you need full comparison:
+;;
+;; - [[clojure.core/=]] compares instances with instance equality
+;; - [[eq]] compares [[TapeCell]] instances by tag, primal and partials
+;;
+;; while:
+;;
+;; - [[g/zero?]], [[g/one?]], [[g/identity?]], [[equiv]] and [[compare]] only
+;;   examine the [[tape-primal]] of either side.
+
+(defn eq
+  "For non-[[TapeCell]]s, identical to [[emmy.value/=]].
+  For [[TapeCell]] instances, equality acts on [[tape-tag]]
+  and [[tape-partials]] too.
+
+  If you want to ignore the tangent components, use [[equiv]]."
+  ([_] true)
+  ([a b]
+   (let [ta? (tape? a) tb? (tape? b)]
+     (cond (and ta? tb?)
+           (let [a ^TapeCell a
+                 b ^TapeCell b]
+             (and (v/= (.-primal a) (.-primal b))
+                  (v/= (.-tag a) (.-tag b))
+                  (v/= (.-in->partial a) (.-in->partial b))))
+
+           ta? (and (empty? (tape-partials a))
+                    (v/= (tape-primal a) b))
+
+           tb? (and (empty? (tape-partials b))
+                    (v/= (tape-primal b) a))
+
+           :else (v/= a b))))
+  ([a b & more]
+   (if (eq a b)
+     (if (next more)
+       (recur b (first more) (next more))
+       (eq b (first more)))
+     false)))
+
+(defn equiv
+  "Returns true if all of the supplied objects have equal [[tape-primal]]s, false
+  otherwise.
+
+  Use [[equiv]] if you want to compare non-[[TapeCell]]s with
+  [[TapeCell]]s and ignore all tangent components. If you _do_ want to take the
+  tangent components into account, prefer [[eq]]."
+  ([_] true)
+  ([a b]
+   (v/= (tape-primal a)
+        (tape-primal b)))
+  ([a b & more]
+   (if (equiv a b)
+     (if (next more)
+       (recur b (first more) (next more))
+       (equiv b (first more)))
+     false)))
+
+(defn compare
+  "Comparator that compares [[Differential]] instances with each other or
+  non-differentials using only the [[finite-part]] of each instance. Matches the
+  response of [[equiv]].
+
+  Acts as [[emmy.value/compare]] for non-[[TapeCell]]s."
+  [a b]
+  (v/compare
```

Comment:
Haskell calls this pattern `on v/compare tape-primal`, usually written infix. I don't know how RH could resist that one after `just`